src/passes/OptimizeInstructions.cpp

```diff
@@ -1117,6 +1117,18 @@ struct OptimizeInstructions
   }
 
   void visitRefEq(RefEq* curr) {
+    // Canonicalize nulls to make the next pattern simpler (and also it may
+    // compress better).
+    if (curr->left->is<RefNull>()) {
+      std::swap(curr->left, curr->right);
+    }
+
+    // RefEq of a value to Null can be replaced with RefIsNull.
+    if (curr->right->is<RefNull>()) {
+      replaceCurrent(Builder(*getModule()).makeRefIs(RefIsNull, curr->left));
+      return;
+    }
```

Comment:
Is there a case when we are comparing to `ref.is_null`s so we can end up with `i32.const 1`? Should we move these after the `areConsecutiveInputsEqual` check below?

Reply:
1. Hmm, at the moment we don't optimize `ref.is_null` to 1 ever - but we should! I'll add that in a followup. In that case the two optimizations would work together in a single pass (a replacement will cause the rules to all be tried again).

2. I don't think the order of the cases matters here, as there is no overlap. I just put this first because it is faster to compute, for lack of a better reason.

Reply:
(for 1, actually there is no need for a followup, as the Precompute pass will turn `ref.is_null(ref.null)` into 1 already)

Reply:
> 2. I don't think the order of the cases matters here, as there is no overlap. I just put this first because it is faster to compute, for lack of a better reason.

If the both sides are `ref.null`, aren't they considered as "ConsecutiveInputsEqual" and can be converted to `i32.const 1` here if we do `areConsecutiveInputsEqual` first? For example,

```wasm
      (ref.eq
        (ref.null eq)
        (ref.null eq)
      )
```
Currently this will be converted to `(ref.is_null (ref.null eq))`, but can't we make this just `(i32.const 1)`?

Maybe doesn't matter much if we do that anyway, but looks simple enought to be done.

Reply:
Oh, you're right, these patterns do collide there...

While precompute will turn it into a constant later anyhow, it makes sense to do it more eagerly here since it's just a reordering.